js-packages/@prestojs/viewmodel/src/useViewModelChangeEffect.ts

```diff
@@ -0,0 +1,174 @@
+     * as the first parameter.
+     */
+    runOnAdd?: boolean;
+    /**
+     * If true then when any records are deleted it will call `run` with 'DELETE'
+     * as the first parameter.
+     */
+    runOnDelete?: boolean;
+    /**
+     * If true then when any records are updated it will call `run` with 'UPDATE'
+     * as the first parameter.
+     */
+    runOnUpdate?: boolean;
+    /** If true will pass through diff object to `run` on second parameter */
+    includeDiff?: boolean;
+};
+
+type AddedDiff<T extends ViewModel> = {
+    previous: T[];
+    added: T[];
+};
+
+type UpdatedDiff<T extends ViewModel> = {
+    previous: T[];
+    updated: T[];
+};
+
+type DeletedDiff<T extends ViewModel> = {
+    previous: T[];
+    deleted: T[];
+};
+
+type Diff<T extends ViewModel> = AddedDiff<T> | UpdatedDiff<T> | DeletedDiff<T>;
+
+function isEqual<T extends ViewModel>(a: T[], b: T[]): boolean {
+    if (a.length !== b.length) {
+        return false;
+    }
+    for (const i in a) {
+        if (!a[i].isEqual(b[i])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Run an effect after a collection of records changes.
+ *
+ * This works by monitoring the collection of records passed in - it does
+ * _not_ listen to changes in a cache. As such make sure you only pass in
+ * latest version of records from a cache (eg. by using useViewModelCache).
+ * If you pass in records that are cached on your component then no changes
+ * in the cache will be detected.
+ *
+ * @param records The records to monitor for changes. This should usually
+ * come from a `useViewModelCache` or another hook that returns data live
+ * from he cache. If a falsy value is passed then no changes will be detected.
+ * Use this to opt out of checks (for example when a fetch is in progress).
+ * @param run The function to call when `records` changes. The first parameter
+ * will be one of 'ADD', 'UPDATE' or 'DELETE' depending on the change. If
+ * `options.includeDiff` is true then the second parameter will be an object
+ * describing the change with the shape:
+ *
+ * ```
+ * {
+ *     // Previous version of `records` before the change
+ *     previous: [...]
+ *     // The record(s) updated
+ *     updated: [...],
+ *     // The record(s) deleted
+ *     deleted: [...],
+ *     // The record(s) added
+ *     added: [...],
+ * }
+ * ```
+ *
+ * One of `updated`, `deleted` or `added` will be set depending on the type
+ * of change.
+ *
+ * Example:
+ *
+ * ```jsx
+ * export default function UserListView() {
+ *   const { data, revalidate, isValidating } = useEndpoint(User.endpoints.list);
+ *   // Refetch data whenever underlying cache changes
+ *   const allRecords = useViewModelCache(User, cache => cache.getAll(fieldList));
+ *   // Note that we pass false while data is being fetched from the backend
+ *   useViewModelChangeEffect(!isValidating && allRecords, revalidate);
+ *   return <ListView records={data} />;
+ * }
```

Comment:
Can we talk about this?

I must have a basic misunderstanding of what's going on here, this is what I understand:

- On the first render
    - useEndpoint() triggers an ajax call; isValidating=true
    - allRecords is empty since nothing is in the cache yet
    - useVieModelChangeEffect does nothing
- Once the ajax call has returned
    - useEndpoint() populates the cache; isValidating=false
    - allRecords is now the full list of fetched records from useEndpoint()
    - useViewModelChangeEffect now ??? (looks like it triggers a revalidate because `records` passed in has changed, but why would we do this?)

Reply:
Can you add a comment somewhere that

- usually you don't want multiple useViewModelChangeEffects on the exact same subset of data as you will trigger redundant ajax queries.
- if you a record is updated & saved elsewhere, then useViewModelCache will return new data which trigger useViewModelChangeEffect. We still want useViewModelChangeEffect () to trigger the useEndpoint() revalidate because that record update may have changed the order of records or caused it to [no longer] appear in a filtered list of data